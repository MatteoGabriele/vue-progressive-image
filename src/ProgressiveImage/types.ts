export type ProgressiveImageProps = {
  src: string;
  placeholderSrc?: string;
  fallbackSrc?: string;
  alt?: string;
  title?: string;
  customClass?: string;
  blur?: number;
  lazyPlaceholder?: boolean;
  objectCover?: boolean;
  delay?: number;
};

export type ProgressiveImageGlobalProps = Omit<
  ProgressiveImageProps,
  "src" | "objectCover | title | alt"
>;
