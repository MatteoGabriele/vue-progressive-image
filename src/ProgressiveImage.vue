<script setup lang="ts">
import useImage from "@/composables/useImage";
import useIntersect from "@/composables/useIntersect";
import {
  MAIN_IMAGE_LOAD_ERROR,
  MAIN_IMAGE_LOAD_SUCCESS,
  globalPropsKey,
} from "@/constants";
import type { ProgressiveImageProps } from "@/types";
import { computed, inject, onMounted, reactive, ref } from "vue";

const props = defineProps<ProgressiveImageProps>();
const globalProps = inject(globalPropsKey, {});
const mergedProps = reactive<ProgressiveImageProps>({
  ...props,
  ...globalProps,
});

const emit = defineEmits([MAIN_IMAGE_LOAD_SUCCESS, MAIN_IMAGE_LOAD_ERROR]);

const rootRef = ref(null);
const imageRef = ref(null);
const isMainImageRendered = ref(false);
const isFallbackImageRendered = ref(false);

const imageSrc = computed<string | undefined>(() => {
  return isFallbackImageRendered.value
    ? mergedProps.fallbackSrc
    : mergedProps.src;
});

const { isIntersected, watchIntersectionOnce } = useIntersect(rootRef);
const { loadImage, aspectRatio, width } = useImage(imageRef);
const isLoading = computed(() => !isMainImageRendered.value);

const paddingHack = computed(() => ({
  paddingBottom: `${aspectRatio.value * 100}%`,
}));

const componentStyle = computed(() => {
  if (mergedProps.objectCover || width.value === 0) {
    return;
  }

  return {
    maxWidth: `${width.value}px`,
  };
});

const imageClasses = computed(() => {
  return [
    mergedProps.customClass,
    {
      "v-progressive-image-object-cover": mergedProps.objectCover,
      "v-progressive-image-loading": isLoading.value,
    },
  ];
});

function toNumber(input: string | number | undefined): number {
  let value = 0;

  if (typeof input === "string") {
    value = Number.parseInt(input);
  } else if (typeof input === "number") {
    value = input;
  }

  return value;
}

function mainImageHandler() {
  loadImage()
    .then(() => {
      setTimeout(() => {
        isMainImageRendered.value = true;
        emit(MAIN_IMAGE_LOAD_SUCCESS);
      }, toNumber(mergedProps.delay));
    })
    .catch((error) => {
      isMainImageRendered.value = true;
      isFallbackImageRendered.value = true;
      emit(MAIN_IMAGE_LOAD_ERROR, error);
    });
}

onMounted(() => {
  if (mergedProps.placeholderSrc && mergedProps.blur) {
    document.documentElement.style.setProperty(
      "--progressive-image-blur",
      `${toNumber(mergedProps.blur)}px`,
    );
  }

  if (mergedProps.src) {
    watchIntersectionOnce(mainImageHandler);
  }
});
</script>

<template>
  <div
    ref="rootRef"
    class="v-progressive-image"
    :class="imageClasses"
    :style="componentStyle"
  >
    <div :style="paddingHack">
      <transition
        :css="!mergedProps.placeholderSrc"
        name="v-progressive-image-main-fade"
        appear
      >
        <img
          v-if="isIntersected"
          v-show="isMainImageRendered"
          ref="imageRef"
          class="v-progressive-image-main"
          :src="imageSrc"
          :alt="mergedProps.alt"
          :title="mergedProps.title"
        />
      </transition>

      <template v-if="mergedProps.placeholderSrc">
        <transition name="v-progressive-image-placeholder-fade" appear>
          <img
            v-if="isLoading"
            class="v-progressive-image-placeholder"
            :loading="mergedProps.lazyPlaceholder ? 'lazy' : 'eager'"
            :src="mergedProps.placeholderSrc"
          />
        </transition>
      </template>

      <div v-if="$slots.default" class="v-progressive-image-slot-default">
        <slot :isLoading="isLoading" />
      </div>
    </div>
  </div>
</template>
