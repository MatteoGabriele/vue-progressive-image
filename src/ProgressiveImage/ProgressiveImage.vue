<script setup>
import { ref, onMounted, computed } from "vue";
import { MAIN_IMAGE_LOAD_SUCCESS, MAIN_IMAGE_LOAD_ERROR } from "@/constants";
import useImage from "@/composables/useImage";
import useIntersect from "@/composables/useIntersect";

const emit = defineEmits([MAIN_IMAGE_LOAD_SUCCESS, MAIN_IMAGE_LOAD_ERROR]);

const props = defineProps({
  src: String,
  placeholderSrc: String,
  fallbackSrc: String,
  alt: String,
  title: String,
  customClass: String,
  blur: [Number, String],
  lazyPlaceholder: {
    type: Boolean,
    default: false,
  },
  delay: {
    type: [Number, String],
    default: 0,
  },
  objectCover: {
    type: Boolean,
    default: false,
  },
});

const rootRef = ref(null);
const imageRef = ref(null);
const isMainImageRendered = ref(false);
const isFallbackImageRendered = ref(false);

const { isIntersected, watchIntersectionOnce } = useIntersect(rootRef);
const { loadImage, aspectRatio, width } = useImage(imageRef);
const isLoading = computed(() => !isMainImageRendered.value);

const paddingHack = computed(() => ({
  paddingBottom: `${aspectRatio.value * 100}%`,
}));

const componentStyle = computed(() => {
  if (props.objectCover || width.value === 0) {
    return;
  }

  return {
    maxWidth: `${width.value}px`,
  };
});

const imageClasses = computed(() => {
  return [
    props.customClass,
    {
      "v-progressive-image-object-cover": props.objectCover,
      "v-progressive-image-loading": isLoading.value,
    },
  ];
});

const mainImageHandler = () => {
  loadImage()
    .then(() => {
      setTimeout(() => {
        isMainImageRendered.value = true;
        emit(MAIN_IMAGE_LOAD_SUCCESS);
      }, props.delay * 1);
    })
    .catch((error) => {
      isMainImageRendered.value = true;
      isFallbackImageRendered.value = true;
      emit(MAIN_IMAGE_LOAD_ERROR, error);
    });
};

onMounted(() => {
  if (props.placeholderSrc && props.blur) {
    document.documentElement.style.setProperty(
      "--progressive-image-blur",
      `${props.blur * 1}px`
    );
  }

  if (props.src) {
    watchIntersectionOnce(mainImageHandler);
  }
});
</script>

<template>
  <div
    ref="rootRef"
    class="v-progressive-image"
    :class="imageClasses"
    :style="componentStyle"
  >
    <div :style="paddingHack">
      <transition
        :css="!placeholderSrc"
        name="v-progressive-image-main-fade"
        appear
      >
        <img
          v-if="isIntersected"
          v-show="isMainImageRendered"
          ref="imageRef"
          class="v-progressive-image-main"
          :src="isFallbackImageRendered ? fallbackSrc : src"
          :alt="alt"
          :title="title"
        />
      </transition>

      <template v-if="placeholderSrc">
        <transition name="v-progressive-image-placeholder-fade" appear>
          <img
            v-if="isLoading"
            class="v-progressive-image-placeholder"
            :loading="lazyPlaceholder ? 'lazy' : 'eager'"
            :src="placeholderSrc"
          />
        </transition>
      </template>

      <div v-if="$slots.default" class="v-progressive-image-slot-default">
        <slot :isLoading="isLoading" />
      </div>
    </div>
  </div>
</template>
