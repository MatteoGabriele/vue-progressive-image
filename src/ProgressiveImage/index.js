import ProgressiveImage from "./ProgressiveImage.vue";

export { ProgressiveImage };

export default ProgressiveImage;
