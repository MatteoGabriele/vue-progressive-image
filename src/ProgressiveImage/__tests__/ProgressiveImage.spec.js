import { mount } from "@vue/test-utils";
import { describe, expect, test, vi } from "vitest";
import ProgressiveImage from "../ProgressiveImage.vue";

vi.mock("../../composables/useIntersect.js", () => {
  return {
    default: () => {
      return {
        isIntersected: () => true,
        watchIntersectionOnce: (fn) => fn(),
      };
    },
  };
});

describe("ProgressiveImage", () => {
  test("render image", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        src: "main-image.jpg",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("render image with placeholder", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        src: "main-image.jpg",
        placeholderSrc: "placeholder-image.jpg",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("use custom blur", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        src: "main-image.jpg",
        placeholderSrc: "placeholder-image.jpg",
        blur: 50,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("use alt attribute", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        src: "main-image.jpg",
        placeholderSrc: "placeholder-image.jpg",
        alt: "image description",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("use object cover", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        src: "main-image.jpg",
        placeholderSrc: "placeholder-image.jpg",
        objectCover: true,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("render default slot", () => {
    const wrapper = mount(ProgressiveImage, {
      slots: {
        default: "<div>lorem ipsum</div>",
      },
      propsData: {
        src: "main-image.jpg",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("render default slot props", () => {
    const wrapper = mount(ProgressiveImage, {
      slots: {
        default: `<template #default="params">{{ params }}</template>`,
      },
      propsData: {
        src: "main-image.jpg",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("lazy load placeholder images", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        placeholderSrc: "placeholder-image.jpg",
        lazyPlaceholder: true,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test("render title attributes", () => {
    const wrapper = mount(ProgressiveImage, {
      propsData: {
        src: "main-image.jpg",
        title: "lorem ipsum dolor sit amet",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
