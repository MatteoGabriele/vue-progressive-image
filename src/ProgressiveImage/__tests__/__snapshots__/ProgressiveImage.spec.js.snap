// Vitest Snapshot v1

exports[`ProgressiveImage > lazy load placeholder images 1`] = `
<div
  class="v-progressive-image v-progressive-image-loading"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <transition-stub>
      <img
        class="v-progressive-image-main"
        style="display: none;"
      />
    </transition-stub>
    <transition-stub>
      <img
        class="v-progressive-image-placeholder"
        loading="lazy"
        src="placeholder-image.jpg"
      />
    </transition-stub>
    <!--v-if-->
  </div>
</div>
`;

exports[`ProgressiveImage > render default slot 1`] = `
<div
  class="v-progressive-image v-progressive-image-loading"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <transition-stub>
      <img
        class="v-progressive-image-main"
        src="main-image.jpg"
        style="display: none;"
      />
    </transition-stub>
    <!--v-if-->
    <div
      class="v-progressive-image-slot-default"
    >
      
      <div>
        lorem ipsum
      </div>
      
    </div>
  </div>
</div>
`;

exports[`ProgressiveImage > render default slot props 1`] = `
<div
  class="v-progressive-image v-progressive-image-loading"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <transition-stub>
      <img
        class="v-progressive-image-main"
        src="main-image.jpg"
        style="display: none;"
      />
    </transition-stub>
    <!--v-if-->
    <div
      class="v-progressive-image-slot-default"
    >
      
      {
  "isLoading": true
}
      
    </div>
  </div>
</div>
`;

exports[`ProgressiveImage > render image 1`] = `
<div
  class="v-progressive-image v-progressive-image-loading"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <transition-stub>
      <img
        class="v-progressive-image-main"
        src="main-image.jpg"
        style="display: none;"
      />
    </transition-stub>
    <!--v-if-->
    <!--v-if-->
  </div>
</div>
`;

exports[`ProgressiveImage > render image with placeholder 1`] = `
<div
  class="v-progressive-image v-progressive-image-loading"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <transition-stub>
      <img
        class="v-progressive-image-main"
        src="main-image.jpg"
        style="display: none;"
      />
    </transition-stub>
    <transition-stub>
      <img
        class="v-progressive-image-placeholder"
        loading="eager"
        src="placeholder-image.jpg"
      />
    </transition-stub>
    <!--v-if-->
  </div>
</div>
`;

exports[`ProgressiveImage > render title attributes 1`] = `
<div
  class="v-progressive-image v-progressive-image-loading"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <transition-stub>
      <img
        class="v-progressive-image-main"
        src="main-image.jpg"
        style="display: none;"
        title="lorem ipsum dolor sit amet"
      />
    </transition-stub>
    <!--v-if-->
    <!--v-if-->
  </div>
</div>
`;

exports[`ProgressiveImage > use alt attribute 1`] = `
<div
  class="v-progressive-image v-progressive-image-loading"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <transition-stub>
      <img
        alt="image description"
        class="v-progressive-image-main"
        src="main-image.jpg"
        style="display: none;"
      />
    </transition-stub>
    <transition-stub>
      <img
        class="v-progressive-image-placeholder"
        loading="eager"
        src="placeholder-image.jpg"
      />
    </transition-stub>
    <!--v-if-->
  </div>
</div>
`;

exports[`ProgressiveImage > use custom blur 1`] = `
<div
  class="v-progressive-image v-progressive-image-loading"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <transition-stub>
      <img
        class="v-progressive-image-main"
        src="main-image.jpg"
        style="display: none;"
      />
    </transition-stub>
    <transition-stub>
      <img
        class="v-progressive-image-placeholder"
        loading="eager"
        src="placeholder-image.jpg"
      />
    </transition-stub>
    <!--v-if-->
  </div>
</div>
`;

exports[`ProgressiveImage > use object cover 1`] = `
<div
  class="v-progressive-image v-progressive-image-object-cover v-progressive-image-loading"
>
  <div
    style="padding-bottom: 56.25%;"
  >
    <transition-stub>
      <img
        class="v-progressive-image-main"
        src="main-image.jpg"
        style="display: none;"
      />
    </transition-stub>
    <transition-stub>
      <img
        class="v-progressive-image-placeholder"
        loading="eager"
        src="placeholder-image.jpg"
      />
    </transition-stub>
    <!--v-if-->
  </div>
</div>
`;
